Add collection_metadata, make metadata immutable

bitdivision opened the pull request:
### What is the context of this PR?
[Trello](https://trello.com/c/ET4X7qU4/2343-eq-survey-runner-1731-startedat-wiped-from-metadata-when-a-questionnaire-is-re-launched-issue)

Closes #1731 

Metadata is passed in and shouldn't be changed. It should be immutable.

We have metadata around the collection of the questionnaire which can't be immutable and doesn't fit in with the rest of `metadata` (started_at). Store it in `collection_metadata`.

`MappingProxyType` is used to make `metadata` a read-only view of `_metadata`. 

### How to review 
- Launch a questionnaire taking note of `RU Ref` and `Collection Exercise SID`
- Tap start survey and ensure `started_at` has been set.
- Launch with the same `RU Ref` and `Collection Exercise SID` and ensure `started_at` doesn't change.

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

bitdivision committed:
Add collection_metadata, immutable metadata
Adds a collection_metadata field to the questionnaire.
collection_metadata should be used for any state around the
collection of a questionnaire where metadata should be
passed in and left unchanged.
Metadata is now an immutable view of the internal _metadata

jonnyshaw89 commented on app/questionnaire/completeness.py:
```diff
@@ -15,13 +15,16 @@ class Completeness:
     COMPLETED_STATES = (COMPLETED, INVALID, SKIPPED)
 
-    def __init__(self, schema, answer_store, completed_blocks, routing_path, metadata):
+    def __init__(self, schema, answer_store, completed_blocks, routing_path, metadata, collection_metadata=None):
```
Why are we passing this into here if we don't use it?

jonnyshaw89 commented on app/submitter/converter.py:
```diff
@@ -58,11 +59,11 @@ def convert_answers(metadata, schema, answer_store, routing_path, flushed=False)
-    if 'case_id' in metadata and metadata['case_id']:
+    if collection_metadata.get('started_at'):
+        payload['started_at'] = collection_metadata['started_at']
+    if metadata.get('case_id'):
```
👍

jonnyshaw89 reviewed (changes requested):
Looks good, have you tested launching before this change then switching to this code in flight?
to mimic a deployment with in-flight users?
also we need to test having this new code and then not. as when deploying both versions of the app are "in" at the same time. therefore someone could have `COLLECTION_METADATA` set on the old code?

jonnyshaw89 commented on app/questionnaire/completeness.py:
```diff
@@ -22,6 +22,9 @@ def __init__(self, schema, answer_store, completed_blocks, routing_path, metadat
 
+        if collection_metadata is None:
+            collection_metadata = {}
+
```
psst :point_up:

bitdivision replied:
The shame...

bitdivision commented:
Very good points about in-flight data.

I've added some logic to pull in the `started_at` time from metadata if it exists. This will need to be removed after deployment. I'll raise a ticket to do that. I've tested that this works by starting a survey on master, switching to this branch, and checking that the payload includes the original started at date.

In terms of starting a questionnaire on this branch and moving back to master, `started_at` also would have been incorrect, since it `metadata` would be used. To avoid this, I've added a couple of nasty bits of code to write started_at to `metadata` as well as `collection_metadata` for the moment. This should all be reverted after deployment.

I've tested moving back to master on the latest version and `started_at` remained set to the original value.

If anyone can see a nicer method of handling in-flight data I'd be all ears.

@jonnyshaw89 Could you take another look?

jonnyshaw89 commented:
@bitdivision looking good, just the codecov now.

bitdivision commented:
Hopefully fixed now. I'll raise a ticket to remove the following code after this is deployed:

- tests/integration/introduction/test_introduction: test_started_at_retrieved_f
rom_metadata_if_available()

- tests/app/submitter/test_converter.py: test_started_at_comes_from_metadata_if
_necessary()

- app/views/questionnaire.py: _set_started_at_metadata_if_required()
Remove indicated lines.

- app/data/model_questionnair_store.py: QuestionnaireStore.set_metadata_key
This method can be removed. It is used to set data on the readonly metadata sto
re.

I suppose the other thing to think about is how long we need to wait after deployment before we can be sure collection_metadata is the proper source of truth.

AndrewTorrance commented on tests/integration/introduction/test_introduction.py:
```diff
@@ -42,3 +43,33 @@ def test_intro_basis_for_completion_not_displayed(self):
+        # Then started_at should be set in collection_metadata (and payload)
+        actual = self.dumpSubmission()['submission']
+
+        self.assertIsNotNone(actual['started_at'])
```
Shouldn't this check if the value of started at is a valid date format , it could be an empty string ?

bitdivision replied:
Yep, I'll sort that out.

AndrewTorrance reviewed (changes requested):
I think there are a few more test cases possibly needed ? Launch a survey twice . We should retain the start date of the first . (not sure if we need variants on this with start date passed in / calculated ?) . Also one test is only testing that start date exists , not that it is a valid date

insacuri reviewed (approved):
Tested it, looks good

insacuri commented on app/views/questionnaire.py:
```diff
@@ -330,13 +332,25 @@ def get_view_submission(schema, eq_id, form_type):  # pylint: disable=unused-arg
+        started_at = datetime.utcnow().isoformat()
+
+        # To avoid loss of in-flight data while collection_metadata is deployed.
+        # After deployment of collection metdata, remove from here to the next comment.
```
Loving this.  Decent instructions on how to remove code after it's not useful anymore is welcome to see!

bitdivision commented:
@AndrewTorrance Could you take a look at the test in test_multiple_login? I think that resolves your point.

If this is okay, I'll squash and sign.

bitdivision commented:
@jonnyshaw89 @dwyeradam Any objections to my merging this?

jonnyshaw89 commented:
@bitdivision yeh, don't merge your own.
